### PR TITLE
[ADD] Added apps Download file creation 2

### PR DESCRIPTION
**Pull request by bizzappdev**
original discussion and approval was done on [Original PR](https://github.com/OCA/apps-store/pull/3)
because of forced push PR was automatically closed. 
So Creating new PR.

### CODE REVIEW

**Comment by elicoidal** [comment]
@gurneyalex @JayVora-SerpentCS @agenterp @bealdav @lasley @astirpe @RoelAdriaans-B-informed 
Sorry to bother you but for extraordinary reasons the initial PR #3 has been lost. 
This new PR has been raised from latest version of the PR. You can still review the remarks in #3 but the files are somehow lost
Would you be so kind to have a final check so that we can merge this one?
Thanks!

**Review by elicoidal — APPROVED** [review]
LGTM

**Review by astirpe — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -15,6 +15,12 @@ addon | version | summary
 --- | --- | ---
 [apps_product_creator](apps_product_creator/) | 10.0.1.0.0 | Create a product variant (by version) for every module found in the Github repositories
 
+
+Unported addons
```
Comment (moylop260) [inline README.md]: This part is updated automatically
Could you revert this change?

File: .travis.yml [hunk 2]
```diff
@@ -22,7 +22,7 @@ env:
   - VERSION="10.0" TESTS="0" LINT_CHECK="0" TRANSIFEX="0"
   - TRANSIFEX_USER='transbot@odoo-community.org'
   - secure: nBOrvOvwQYVtKAt6VNVcpKkLVDMSHFjRHlfOQPEF5tZUb6ck7cmusML8/NkUgQSw6N1PmkBE2KLfw6PHHZd6QTWQuDTou6CjstyWQbJH/tX6DL3jdAvTh8Z+VyRxZeKRT7qnCqhs47z+2v/JP/4tUlWLsube4AO5m0g8D3rRllf8zMMg+zG2kf57a7FCp1XeoBWGu30rQLI2US8x/dtxNpb4wj3XoiJkXq5hj6grmJG+b1io8nmfW4AKDTf3A0MvE43hNtfSDAEcGf9g9K+zRPSrdj/7qU/QIXpBRIDMQrW3kw/h4+DdPa0bZDe8uStgXipNw6M1gUwD6B07OgqV5hwM9LowfbMxh+H7QtVw6jJoMrQuC+rV25ldo6jzYqjfgTg4LIDcdZjc3AQ5MVmfWbzl1gczCaTTMiHz0JYTk+9uuzNYZghY9ZEj2e1D3nTHeQOqlZQJ1d0N17KN6661rK2gqlzP+bGTszLpEKIFN5ST2mu9WgiwADYmdmnx7ywG9/TZOHY3XIYOqYoTG+8ZeDRl2msr57Ij4M3nvzZbkOmurLe/0lz4qH3ThfkikSccmxqvlKWi85tkE4Eo+sh/ilfGAYt0ClJGImtaCFGvD1Oo4+LBFfQCyX+NCh4gtnhwDps3Z8IUtClW/5kDf6bD8jb6jJ+Nrs2Tp4vCLsT7T1I=
-  
+
```
Comment (moylop260) [inline .travis.yml]: Unnecessary change for this PR

File: apps_download/models/product_product.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'product_product_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
```
Comment (moylop260) [inline apps_download/models/product_product.py]: If the string is the same that the name of the field is not necessary define the string because is used by default.

3rd point from https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#fields

File: apps_download/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                _('Error: You cannot create recursive dependency.')
+            )
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
+                continue
+            res += child.dependent_product_ids
```
Comment (moylop260) [inline apps_download/models/product_product.py]: What about use `res |=` instead of `+=`?

File: apps_download/models/product_product.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        if not self._check_m2m_recursion('dependent_product_ids'):
+            raise ValidationError(
+                _('Error: You cannot create recursive dependency.')
+            )
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
```
Comment (moylop260) [inline apps_download/models/product_product.py:37]: You could avoid this addtional `if` since that `res |= []` concatenate nothing

File: apps_download/models/product_product.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            res += self.child_dependency(child.dependent_product_ids)
+        return res
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += self.child_dependency(
```
Comment (moylop260) [inline apps_download/models/product_product.py]: What about `|=`?

File: apps_download/models/product_product.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
```
Comment (moylop260) [inline apps_download/models/product_product.py]: What about `for product in self.filtered('module_path'):`?

File: apps_download/models/product_product.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            if not product.module_path:
+                continue
+            tmp_dir = tempfile.mkdtemp()
+            tmp_dir_2 = tempfile.mkdtemp()
+            dependent_products = product.create_dependency_list()
+            dependent_products = dependent_products[product.id]
+            for dependent_product in dependent_products:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
```
Comment (moylop260) [inline apps_download/models/product_product.py]: What about use a python library in order to compatibility with multi-platform? (Maybe `shutil`)

File: apps_download/models/product_product.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                        'datas': data_encode,
+                        'datas_fname':  (product.name + time_version_value +
+                                         '.zip'),
+                        'type': 'binary',
+                        'name': product.name + time_version_value + '.zip',
+                        'res_model': product._name,
+                        'res_id': product.id,
+                        'product_downloadable': True,
+                    })
+                except:
```
Comment (moylop260) [inline apps_download/models/product_product.py]: What about catch the exception to add the message of the error in the logger?

**Review by moylop260 — CHANGES_REQUESTED** [review]

**Comment by elicoidal** [comment]
Thanks @moylop260 for the review!

File: apps_download/models/product_product.py (hunk 5 above)
Comment (bizzappdev) [reply apps_download/models/product_product.py:37]: added |= but IMO if condition will be better so that at least the function will not be called which is not needed in case if there are no dependent products.

**Comment by bizzappdev** [comment]
@moylop260 all other things are taken care please review again.

**Comment by elicoidal** [comment]
@bizzappdev thanks!
By the way can you help me to get the last PR open for the frontend?
I remember that you had the contact information. Is it @agenterp?

**Comment by bizzappdev** [comment]
@elicoidal  yes he was from @agenterp. 
after this pr accepted we have to plan 2 things.
1. integrate apps_download with apps_product_creator module. (I can take care)
2. frontend module. (I and @agenterp together can take care)

**Comment by elicoidal** [comment]
@moylop260 Can you have a look?

**Comment by elicoidal** [comment]
@bizzappdev Can we start the last piece of PR?

**Review by moylop260 — APPROVED** [review]

**Comment by elicoidal** [comment]
Thanks @moylop260 !

**Comment by elicoidal** [comment]
@bizzappdev @agenterp let's move forward with the last piece.
It would be good to have at least a WIP PR so that I could leverage effort in the community.
Looking forward to reading you

**Comment by bizzappdev** [comment]
@elicoidal  okay sure. I will arrange the new branch and start working on that. @agenterp can you somehow provide us with the website view code which we created during Code sprint.

**Comment by bizzappdev** [comment]
@elicoidal  we have one bad news somehow we do not have the design of HTML/XML page created on oca code sprint.
Does anyone have the design on paper or something from which we can create from scratch?

**Comment by elicoidal** [comment]
@bizzappdev Unfortunately not in my hand: I thought it was been done. Can we have some screenshot at all of the previous work?
I can start a design but basically we should simply replicate the Odoo apps page structure for starter then we can improve.
Otherwise I can work on it this week end
